File: ffi/bindings/java/examples/src/main/java/io/stepfunc/rodbus/examples/ClientTlsExample.java
package io.stepfunc.rodbus.examples;

import static org.joou.Unsigned.*;

import io.stepfunc.rodbus.*;
import io.stepfunc.rodbus.Runtime;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;

public class ClientTlsExample {
    public static void main(String[] args) throws Exception {
        // ANCHOR: logging_init
        // initialize logging with the default configuration
        Logging.configure(new LoggingConfig(), new ConsoleLogger());
        // ANCHOR_END: logging_init

        // initialize the runtime
        // ANCHOR: runtime_init
        RuntimeConfig runtimeConfig = new RuntimeConfig();
        runtimeConfig.numCoreThreads = ushort(4);
        Runtime runtime = new Runtime(runtimeConfig);
        // ANCHOR_END: runtime_init

        // ANCHOR: tls_self_signed_config
        TlsClientConfig selfSignedTlsConfig =
            new TlsClientConfig(
                "test.com",
                "./certs/self_signed/entity2_cert.pem",
                "./certs/self_signed/entity1_cert.pem",
                "./certs/self_signed/entity1_key.pem",
                "" // no password
        );
        selfSignedTlsConfig.certificateMode = CertificateMode.SELF_SIGNED;
        // ANCHOR_END: tls_self_signed_config

        // ANCHOR: tls_ca_chain_config
        TlsClientConfig caChainTlsConfig =
            new TlsClientConfig(
                "test.com",
                "./certs/ca_chain/ca_cert.pem",
                "./certs/ca_chain/entity1_cert.pem",
                "./certs/ca_chain/entity1_key.pem",
                "" // no password
        );
        // ANCHOR_END: tls_ca_chain_config

        TlsClientConfig tlsConfig = caChainTlsConfig;

        // initialize a Modbus TLS client channel
        // ANCHOR: create_tls_channel
        DecodeLevel decodeLevel = new DecodeLevel();
        Channel channel = Channel.createTlsClient(runtime, "127.0.0.1:802", ushort(100), new RetryStrategy(), tlsConfig, decodeLevel);
        // ANCHOR_END: create_tls_channel

        try {
            run(channel);
        } finally {
            // ANCHOR: runtime_shutdown
            runtime.shutdown();
            // ANCHOR_END: runtime_shutdown
        }
    }

    private static void run(Channel channel) throws Exception {
        // ANCHOR: request_param
        final RequestParam param = new RequestParam(ubyte(1), uint(1000));
        // ANCHOR_END: request_param
        // ANCHOR: address_range
        final AddressRange range = new AddressRange(ushort(0), ushort(5));
        // ANCHOR_END: address_range

        // Handle user input
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            String line = reader.readLine();
            switch (line) {
                case "x":
                    return;
                case "rc": {
                    // ANCHOR: read_coils
                    channel.readCoils(param, range).thenAccept(ClientTlsExample::handleBitResult);
                    // ANCHOR_END: read_coils
                    break;
                }
                case "rdi": {
                    channel.readDiscreteInputs(param, range).thenAccept(ClientTlsExample::handleBitResult);
                    break;
                }
                case "rhr": {
                    channel.readHoldingRegisters(param, range).thenAccept(ClientTlsExample::handleRegisterResult);
                    break;
                }
                case "rir": {
                    channel.readInputRegisters(param, range).thenAccept(ClientTlsExample::handleRegisterResult);
                    break;
                }
                case "wsc": {
                    /// ANCHOR: write_single_coil
                    channel.writeSingleCoil(param, new Bit(ushort(0), true)).thenAccept(ClientTlsExample::handleWriteResult);
                    /// ANCHOR_END: write_single_coil
                    break;
                }
                case "wsr": {
                    channel.writeSingleRegister(param, new Register(ushort(0), ushort(76))).thenAccept(ClientTlsExample::handleWriteResult);
                    break;
                }
                case "wmc": {
                    channel.writeMultipleCoils(param, ushort(0), Arrays.asList(true, false)).thenAccept(ClientTlsExample::handleWriteResult);
                    break;
                }
                case "wmr": {
                    // ANCHOR: write_multiple_registers
                    channel.writeMultipleRegisters(param, ushort(0), Arrays.asList(ushort(0xCA), ushort(0xFE))).thenAccept(ClientTlsExample::handleWriteResult);
                    // ANCHOR_END: write_multiple_registers
                    break;
                }
                default:
                    System.out.println("Unknown command");
                    break;
            }
        }
    }

    // ANCHOR: handle_bit_result
    private static void handleBitResult(BitReadResult result) {
        if (result.result.summary == Status.OK) {
            System.out.println("success!");
            for(Bit bit : result.iterator) {
                System.out.println("index: " + bit.index + " value: " + bit.value);
            }
        } else if (result.result.summary == Status.EXCEPTION) {
            System.out.println("Modbus exception: " + result.result.exception);
        } else {
            System.out.println("error: " + result.result.summary);
        }
    }
    // ANCHOR_END: handle_bit_result

    private static void handleRegisterResult(RegisterReadResult result) {
        // ANCHOR: error_handling
        if (result.result.summary == Status.OK) {
            System.out.println("success!");
            for(Register register : result.iterator) {
                System.out.println("index: " + register.index + " value: " + register.value);
            }
        } else if (result.result.summary == Status.EXCEPTION) {
            System.out.println("Modbus exception: " + result.result.exception);
        } else {
            System.out.println("error: " + result.result.summary);
        }
        // ANCHOR_END: error_handling
    }

    // ANCHOR: handle_write_result
    private static void handleWriteResult(ErrorInfo result) {
        if (result.summary == Status.OK) {
            System.out.println("success!");
        } else if (result.summary == Status.EXCEPTION) {
            System.out.println("Modbus exception: " + result.exception);
        } else {
            System.out.println("error: " + result.summary);
        }
    }
    /// ANCHOR_END: handle_write_result
}

File: ffi/bindings/c/client_tls_example.cpp
#include "rodbus.hpp"

#include <chrono>
#include <iostream>
#include <string>

class Logger : public rodbus::Logger {
    void on_message(rodbus::LogLevel level, std::string message) override
    {
        std::cout << message;
    }
};

// ANCHOR: bit_read_callback
class BitReadCallback : public rodbus::BitReadCallback
{
    void on_complete(rodbus::BitValueIterator& it) override
    {
        std::cout << "success!" << std::endl;
        while (it.next()) {
            const auto value = it.get();
            std::cout << "index: " << value.index << "value: " << value.value << std::endl;
        }
    }
    void on_failure(rodbus::RequestError err) override
    {
        std::cout << "error: " << rodbus::to_string(err) << std::endl;
    }
};
// ANCHOR_END: bit_read_callback

class RegisterReadCallback : public rodbus::RegisterReadCallback
{
    void on_complete(rodbus::RegisterValueIterator& it) override
    {
        std::cout << "success!" << std::endl;
        while (it.next()) {
            const auto value = it.get();
            std::cout << "index: " << value.index << "value: " << value.value << std::endl;
        }
    }
    void on_failure(rodbus::RequestError err) override
    {
        std::cout << "error: " << rodbus::to_string(err) << std::endl;
    }
};

class WriteCallback : public rodbus::WriteCallback
{
    void on_complete(rodbus::Nothing result) override
    {
        std::cout << "success!" << std::endl;
    }
    void on_failure(rodbus::RequestError err) override
    {
        std::cout << "error: " << rodbus::to_string(err) << std::endl;
    }
};

int main()
{
    // ANCHOR: logging_init
    // initialize logging with the default configuration
    rodbus::Logging::configure(rodbus::LoggingConfig(), std::make_unique<Logger>());
    // ANCHOR_END: logging_init

    // initialize the runtime
    // ANCHOR: runtime_create
    auto runtime_config = rodbus::RuntimeConfig();
    runtime_config.num_core_threads = 4;
    auto runtime = rodbus::Runtime(runtime_config);
    // ANCHOR_END: runtime_create

    // ANCHOR: tls_self_signed_config
    auto self_signed_tls_config = rodbus::TlsClientConfig(
        "test.com",
        "./certs/self_signed/entity2_cert.pem",
        "./certs/self_signed/entity1_cert.pem",
        "./certs/self_signed/entity1_key.pem",
        "" // no password
    );
    self_signed_tls_config.certificate_mode = rodbus::CertificateMode::self_signed;
    // ANCHOR_END: tls_self_signed_config

    // ANCHOR: tls_ca_chain_config
    auto ca_chain_tls_config = rodbus::TlsClientConfig(
        "test.com",
        "./certs/ca_chain/ca_cert.pem",
        "./certs/ca_chain/entity1_cert.pem",
        "./certs/ca_chain/entity1_key.pem",
        "" // no password
    );
    // ANCHOR_END: tls_ca_chain_config

    const auto tls_config = ca_chain_tls_config;

    // initialize a Modbus TCP client channel
    // ANCHOR: create_tcp_channel
    auto channel = rodbus::ClientChannel::create_tls(
        runtime,
        "127.0.0.1:802",
        100,
        rodbus::RetryStrategy(),
        tls_config,
        rodbus::DecodeLevel()
    );
    // ANCHOR_END: create_tcp_channel

    // request param that we will be reusing
    // ANCHOR: request_param
    const auto param = rodbus::RequestParam(1, // Unit ID
                                            std::chrono::seconds(1) // Timeout
    );
    // ANCHOR_END: request_param

    // address range that we will be reusing
    // ANCHOR: address_range
    const auto range = rodbus::AddressRange(0, // start
                                            5  // count
    );
    // ANCHOR_END: address_range

    while (true)
    {
        std::string cmd;
        std::getline(std::cin, cmd);

        if (cmd == "x") {
            return 0;
        }
        else if (cmd == "rc") {
            channel.read_coils(param, range, std::make_unique<BitReadCallback>());
        }
        else if (cmd == "rdi") {
            channel.read_discrete_inputs(param, range, std::make_unique<BitReadCallback>());
        }
        else if (cmd == "rhr") {
            channel.read_holding_registers(param, range, std::make_unique<RegisterReadCallback>());
        }
        else if (cmd == "rir") {
            channel.read_input_registers(param, range, std::make_unique<RegisterReadCallback>());
        }
        else if (cmd == "wsc") {
            const auto bit_value = rodbus::BitValue(0, true);
            channel.write_single_coil(param, bit_value, std::make_unique<WriteCallback>());
        }
        else if (cmd == "wsr") {
            const auto register_value = rodbus::RegisterValue(0, 76);
            channel.write_single_register(param, register_value, std::make_unique<WriteCallback>());
        }
        else if (cmd == "wmc") {
            // create the bitlist
            auto bit_list = std::vector<bool>();
            bit_list.emplace_back(true);
            bit_list.emplace_back(false);

            // send the request
            channel.write_multiple_coils(param, 0, bit_list, std::make_unique<WriteCallback>());
        }
        else if (cmd == "wmr") {
            // create the register list
            // ANCHOR: write_multiple_registers
            auto register_list = std::vector<uint16_t>();
            register_list.emplace_back(0xCA);
            register_list.emplace_back(0xFE);

            // send the request
            channel.write_multiple_registers(param, 0, register_list, std::make_unique<WriteCallback>());
            // ANCHOR_END: write_multiple_registers
        }
        else {
            std::cout << "unknown command: " << cmd << std::endl;
        }
    }
}
